src/py/flwr/simulation/ray_transport/ray_actor.py

```diff
@@ -0,0 +1,388 @@
+    ):
+        self.client_resources = client_resources
+        self.create_actor_fn = create_actor_fn
+
+        if actor_list is None:
+            # Figure out how many actors can be created given the cluster resources
+            # and the resources the user indicates each VirtualClient will need.
+            num_actors = pool_size_from_resources(client_resources)
+            actors = [create_actor_fn() for _ in range(num_actors)]
+        else:
+            # When __reduce__ is executed, we don't want to created
+            # a new list of actors again.
+            actors = actor_list
+
+        super().__init__(actors)
+
+        # A dict that maps cid to another dict containing: a reference to the remote job
+        # and its status (i.e. whether it is ready or not)
+        self._cid_to_future: Dict[
+            str, Dict[str, Union[bool, Optional[ObjectRef[Any]]]]
+        ] = {}
+        self.actor_to_remove: Set[str] = set()  # a set
+        self.num_actors = len(actors)
+
+        self.lock = threading.RLock()
+
+    def __reduce__(self):  # type: ignore
+        """Make this class serializable (needed due to lock)."""
+        return VirtualClientEngineActorPool, (
+            self.create_actor_fn,
+            self.client_resources,
+            self._idle_actors,  # Pass existing actors to avoid killing/re-creating
+        )
+
+    def add_actors_to_pool(self, num_actors: int) -> None:
+        """Add actors to the pool.
+
+        This expands the pool after it has been created iif new resources are added to
+        your Ray cluster (e.g. you add a new node).
+        """
+        with self.lock:
+            new_actors = [self.create_actor_fn() for _ in range(num_actors)]
+            self._idle_actors.extend(new_actors)
+            self.num_actors += num_actors
+
+    def submit(self, fn: Any, value: Tuple[Callable[[], ClientRes], str]) -> None:
+        """Take idle actor and assign it a client workload.
+
+        Submit a job to an actor by first removing it from the list of idle actors, then
+        check if this actor was flagged to be removed from the pool
+        """
+        job_fn, cid = value
+        actor = self._idle_actors.pop()
+        if self._check_and_remove_actor_from_pool(actor):
+            future = fn(actor, job_fn)
+            future_key = tuple(future) if isinstance(future, List) else future
+            self._future_to_actor[future_key] = (self._next_task_index, actor, cid)
+            self._next_task_index += 1
+
+            # update with future
+            self._cid_to_future[cid]["future"] = future_key
+
+    def submit_client_job(
+        self, actor_fn: Any, job: Tuple[Callable[[], ClientRes], str]
+    ) -> None:
+        """Submit a job while tracking client ids."""
+        # We need to put this behind a lock since .submit() involves
+        # removing and adding elements from a dictionary. Which creates
+        # issues in multi-threaded settings
+
+        _, cid = job
```

Comment:
```suggestion
        _, cid = job

        # We need to put this behind a lock since .submit() involves
        # removing and adding elements from a dictionary. Which creates
        # issues in multi-threaded settings
```